src/Hyprsunset.cpp

```diff
@@ -147,15 +154,98 @@ int CHyprsunset::init() {
 
 void CHyprsunset::tick() {
     if (g_pIPCSocket && g_pIPCSocket->mainThreadParseRequest()) {
-        // Reload
-        calculateMatrix();
+        reload();
+    }
+}
+
+void CHyprsunset::loadCurrentProfile() {
+    profiles = g_pConfigManager->getSunsetProfiles();
+
+    Debug::log(NONE, "┣ Loaded {} profiles, applying the current one", profiles.size());
+
+    std::sort(profiles.begin(), profiles.end(), [](const auto& a, const auto& b) {
+        if (a.time.hour < b.time.hour)
+            return true;
+        else if (a.time.hour > b.time.hour)
+            return false;
+        else
+            return a.time.minute < b.time.minute;
+    });
+
+    int current = g_pHyprsunset->currentProfile();
 
-        for (auto& o : state.outputs) {
-            o->applyCTM(&state);
+    if (current == -1)
+        return;
+
+    SSunsetProfile profile = g_pHyprsunset->profiles[current];
+    KELVIN                 = profile.temperature;
+    GAMMA                  = profile.gamma;
+    identity               = profile.identity;
+    MAX_GAMMA              = g_pConfigManager->getMaxGamma();
+}
+
+int CHyprsunset::currentProfile() {
+    if (profiles.empty())
+        return -1;
+    else if (profiles.size() == 1)
+        return 0;
+
+    auto now = std::chrono::zoned_time(std::chrono::current_zone(), std::chrono::system_clock::now()).get_local_time();
+
+    for (size_t i = 0; i < profiles.size(); ++i) {
+        const auto& p = profiles[i];
+
+        auto        time = std::chrono::floor<std::chrono::days>(now) + p.time.hour + p.time.minute;
+
+        if (time >= now) {
+            if (i == 0)
+                return profiles.size() - 1;
+            return i - 1;
         }
+    }
 
-        commitCTMs();
+    return 0;
+}
+
+void CHyprsunset::schedule() {
+    std::thread([&]() {
+        while (true) {
+            int current = currentProfile();
+            if (current == -1)
+                break;
+
+            SSunsetProfile nextSettings = (size_t)current == profiles.size() - 1 ? profiles[0] : profiles[current + 1];
+
+            auto           now  = std::chrono::zoned_time(std::chrono::current_zone(), std::chrono::system_clock::now()).get_local_time();
+            auto           time = std::chrono::floor<std::chrono::days>(now) + nextSettings.time.hour + nextSettings.time.minute;
+
+            if (now >= time) {
+                time += std::chrono::days(1);
+            }
+
+            auto system_time = std::chrono::zoned_time{std::chrono::current_zone(), time}.get_sys_time();
+
+            std::this_thread::sleep_until(system_time);
+
+            KELVIN   = nextSettings.temperature;
+            GAMMA    = nextSettings.gamma;
+            identity = nextSettings.identity;
 
-        wl_display_flush(state.wlDisplay);
+            Debug::log(NONE, "┣ Switched to new profile at: {}", time);
+
+            reload();
```

Comment:
the wayland event loop is not thread-safe... Please see https://github.com/hyprwm/xdg-desktop-portal-hyprland/blob/master/src/core/PortalManager.cpp#L295 for an example of how to tackle multi-threading in a wl event loop